spec/request/csrf_request_spec.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'CSRF scenarios', type: :request do
+  # ActionController::Base.allow_forgery_protection = false in the 'test' environment
+  # We explicity enable it for this spec
+  before(:all) do
+    Settings.sentry.dsn = 'truthy'
+    @original_val = ActionController::Base.allow_forgery_protection
+    ActionController::Base.allow_forgery_protection = true
+  end
+
+  after(:all) do
+    Settings.sentry.dsn = nil
+    ActionController::Base.allow_forgery_protection = @original_val
+  end
+
+  before do
+    # innocuous route chosen for setting the CSRF token in the response header
+    get(v0_maintenance_windows_path)
+    @token = response.headers['X-CSRF-Token']
+  end
+
+  describe 'CSRF protection' do
+    before(:all) do
+      Rails.application.routes.draw do
+        namespace :v0, defaults: { format: 'json' } do
+          resources :maintenance_windows, only: [:index]
+        end
+        match 'csrf_test', to: 'v0/example#index', via: :all
+      end
+    end
+
+    after(:all) do
+      Rails.application.reload_routes!
+    end
+
+    %i[post put patch delete].each do |verb|
+      context "for #{verb.upcase} requests" do
+        context 'without a CSRF token present' do
+          it 'raises an exception' do
+            expect(Raven).to receive(:capture_message).with('Request susceptible to CSRF', level: 'info')
+            send(verb, '/csrf_test')
+            # expect(response.status).to eq 500
+          end
+        end
+
+        context 'with a CSRF token present' do
+          it 'succeeds' do
+            send(verb, '/csrf_test', headers: { 'X-CSRF-Token' => @token })
+            expect(response.status).to eq 200
+          end
+        end
+      end
+    end
+
+    context 'for GET requests' do
+      context 'without a CSRF token present' do
+        it 'succeeds' do
+          expect(Raven).not_to receive(:capture_message)
+          get '/csrf_test'
+          expect(response.status).to eq 200
+        end
+      end
+
+      context 'with a CSRF token present' do
+        it 'succeeds' do
+          expect(Raven).not_to receive(:capture_message)
+          get '/csrf_test', headers: { 'X-CSRF-Token' => @token }
+          expect(response.status).to eq 200
+        end
+      end
+    end
+  end
+
+  # SAML callback
+  describe 'POST SAML callback' do
+    context 'without a CSRF token' do
+      context 'v0' do
+        it 'does not raise an error' do
+          expect(Raven).not_to receive(:capture_message).with('Request susceptible to CSRF', level: 'info')
+          post(auth_saml_callback_path)
+          # expect(response.body).not_to match(/ActionController::InvalidAuthenticityToken/)
```

Comment:
What's the purpose of this commented expectation?

Reply:
When we roll this out it will be only logging to Sentry for a bit (a few days?) so we can see if any routes aren't being covered properly by CSRF protection.  Then we'll switch it to actually raise the exception.  At that point I'm going to change these specs.  i.e. remove the `Raven` expectations and start checking that the exception is being raised.